File: python/demo_monads.py
from typing import List, Dict

def square(x: int) -> Dict[str, object]:
    return {
        "result": x * x,
        "logs": [f"Squared {x} to get {x * x}."]
    }

def add_one(x: Dict[str, object]) -> Dict[str, object]:
    result = x["result"] + 1
    logs: List[str] = x["logs"] + [f"Added 1 to {x['result']} to get {result}."]
    return {"result": result, "logs": logs}

out = add_one(square(2))
print(out)

File: scala/DemoMonads.scala
object ScalaDemoMonads extends App {
  case class NumberWithLogs(result: Int, logs: List[String])

  def square(x: Int): NumberWithLogs =
    NumberWithLogs(x * x, List(s"Squared $x to get ${x * x}."))

  def addOne(x: NumberWithLogs): NumberWithLogs =
    NumberWithLogs(x.result + 1,
      x.logs :+ s"Added 1 to ${x.result} to get ${x.result + 1}.")

  val out = addOne(square(2))
  println(out)
}